lib/sem182/msem182.3
.\" 
.\" Modification history:
.\" 	2.1	Changed semantics of mP and mV
.\" 	1.1	Derived from sem182.3
.\"                                                                            *
.TH MSEM182 3 "RELEASE 2.4" "" "The sem182 library"
.SH NAME
mseminit, msemgrab, mP, mV, msemrm \- Semaphore field operations
.SH SYNTAX
.nf
.B #include <msem182.h>
.LP
.B "int mseminit (key_t key, int semperm, int nsems, ...);"
.B "int msemgrab (key_t key, int nsems);"
.B "int mP (int semid, int nsems, ...);"
.B "int mV (int semid, int nsems, ...);"
.B "int msemrm (int semid);"
.fi
.SH DESCRIPTION
These functions are easier to use than the standard interface.
.LP
.I mseminit
creates a semaphore field of 
.I nsems 
semaphores with 
.I key
using
.I semperm
as permissions and initializes the semaphore field to the specified
values.  On success this function returns a valid descriptor for the
semaphore field.  In case a semaphore field with the specified key already exists,
.I mseminit
returns -1.
.LP
.I msemgrab
delivers a valid descriptor for an 
.B already existing
semaphore field of at least 
.I nsems
semaphores.
.LP
.I mP
and
.I mV
perform a
.B P
resp.
.B V
operation on the 
semaphores indicated by the following
.I nsems 
arguments in the field
.I semid
atomically. Thus 
.EX
mP (semid, 3, 1, 0, 2)
.EE
is the same as 
.EX
P (semid1);
P (semid0);
P (semid2);
.EE
except that 
.B mP 
is an atomic action, whereas the three
\fBP\fR's
are not. Using the same index twice in a single call to 
.B mP
or
.B mV
results in undefined behaviour.
.LP
.I msemrm
removes the semaphore field given by
.I semid.
.SH DIAGNOSTICS
All functions return -1 in case of error and set 
.B errno 
accordingly.
.SH COMPILING
You have to specify the library to the compiler, e.g.
.LP
c89 program.c 
.B -lsem182
.SH SEE ALSO
semget(2), semop(2), semctl(2), sem182(3)
.SH DISCLAIMER
The sem182 library is provided "as is" for education purposes      
only. There is no warranty for the correct functionality and we do not          
accept any responsibility whatsoever for any loss or damage that is caused      
by the use of the sem182 library.
.SH AUTHORS
Peter Holzer, by modifying an old version of sem182 by Guenther Leber and
Heinz Kantz, which had also been modified (the version, not Heinz) by
Gustav Pospischil. Additional code contributed by Raimund Kirner.
.LP
Manual page written by Peter Holzer, by modifying a quite recent version
of the manual page of sem182 by Guenther Leber. Manual page clean-up and
reformatting by Thomas Perl.


lib/sem182/msem182.c
/**
 * Copyright (c) 2002, Vienna University of Technology, Real-Time Systems Group.
 * All rights reserved.
 *
 * Contact: sem182@vmars.tuwien.ac.at
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 **/

char msem182_c_sccs_id [] = 
	"@(#)	msem182.c	2.3	2002/03/05	rk";
/*
 * Purpose: 	Convenience functions for Semaphores
 *
 * Motivation:	The semop function is hard to use and provides
 *              functionality which is seldom used or even annoying
 *		(like the fact that 0 isn't a nop). These functions
 *		should provide a cleaner interface.
 *
 * Modification history:
 *  2.5 Use of semctl with enum semun
 *	2.2	Changed order of parameters in mseminit.
 *	2.1	Changed the semantics of mP and mV. The arguments now
 *		specify the index of the semaphore to operate on, 
 *		not the operation.
 *	1.1	got the version from the solution to Ex. 2.11 
 *		(which was derived from an out-of-date version of 
 *		sem182.c which had been heavily hacked by gusti) and
 *		adapted it to the STBS and the newest version of
 *		sem182.c.
 */
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/sem.h>
#include <stdarg.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include "msem182.h"

/*** function mseminit
 * 
 * Purpose: Create and initialize a semaphore field.
 *	    The SF must not exist yet.
 * In:      key Key of semaphore field
 *          nsems   Number of semaphores the field shall contain
 *          perm    permission bits for ugo
 *	    ...	    nsems initial values for semaphores
 * Returns: Semaphore descriptor or -1 in case of error.
 */

int mseminit( key_t key, int perm, int nsems, ...)
{
#if !defined(__GNU_LIBRARY__) || defined(_SEM_SEMUN_UNDEFINED)
     union semun {
         int val;                    /* value for SETVAL */
         struct semid_ds *buf;       /* buffer for IPC_STAT, IPC_SET */
         unsigned short int *array;  /* array for GETALL, SETALL */
     };
#endif

    int     semid;
    union semun semarg;
    va_list ap;
    int     i, initval;

    if (
        (semid = semget (key, nsems, IPC_CREAT | IPC_EXCL | (perm & 0666)))

        < 0
       )
    {
        return -1;
    }

    /* Initialize Semaphore */
    va_start (ap, nsems);
    for (i = 0; i < nsems; i ++) 
    {
        semarg.val = va_arg (ap, int);
        if (semctl (semid, i, SETVAL, semarg) < 0) 
        {
           int save_errno = errno;
           (void) msemrm(semid);
           errno = save_errno;
           return -1;
        }
    }
    va_end (ap);
    
    return semid;
}


/*** function msemgrab
 *
 * Purpose:	grab existing semaphore field
 * In:		key	Key of SF.
 *		nsems	number of semaphores in the SF
 * Returns:	semid or -1 in case of error
 */
int msemgrab (key_t key, int nsems)
{
	return semget (key, nsems, 0);
}


/*** function msemrm
 * 
 * Purpose: remove semaphore field
 * In:	    semid	id of SF to remove
 */
int msemrm(int semid) 
{
    return semctl (semid, 0, IPC_RMID, 0);
}


/*** function mV
 *
 * Purpose: Dijkstra's V for a whole semaphore field.
 * In:	    semid	Id of SF
 *	    nsems	Number of Semaphores in the field.
 *			(I am not sure if I should check this. 
 *			It may help catch some errors, but too many
 *			are caught by the OS, and a smaller number 
 *			might be handy if you want only a work on a
 *			subset).
 *	    ...		nsems values indicating the semaphores on which
 *			a V operation should be performed.
 * Returns: -1 in case of error, else 0.
 */
int mV(int semid, int nsems, ...)
{  
    struct sembuf   *semp;
    int             i;
    va_list         ap;

    if ((semp = calloc (nsems, sizeof (struct sembuf))) == NULL)
    {
        return -1;
    }

    va_start (ap, nsems);
    for (i = 0; i < nsems; i ++) 
    {
        semp[i].sem_num = va_arg (ap, int);
        semp[i].sem_op = 1;
    }
    va_end (ap);

    if (semop(semid, semp, nsems) < 0)
    {
        free (semp);
        return -1;
    }
    free (semp);
    return 0;
}


/*** function mP
 *
 * Purpose: Dijkstra's P for a whole semaphore field.
 * In:	    semid	Id of SF
 *	    nsems	Number of Semaphores in the field.
 *			(I am not sure if I should check this. 
 *			It may help catch some errors, but too many
 *			are caught by the OS, and a smaller number 
 *			might be handy if you want only a work on a
 *			subset).
 *	    ...		nsems values indicating the semaphores on which
 *			a P operation should be performed.
 * Returns: -1 in case of error, else 0.
 */
int mP(int semid, int nsems, ...)
{  
    struct sembuf   *semp;
    int             i;
    va_list         ap;

    if ((semp = calloc (nsems, sizeof (struct sembuf))) == NULL)
    {
        return -1;
    }

    va_start (ap, nsems);
    for (i = 0; i < nsems; i ++) 
    {
        semp[i].sem_num = va_arg (ap, int);
        semp[i].sem_op = -1;
    }
    va_end (ap);

    if (semop(semid, semp, nsems) < 0)
    {
        free (semp);
        return -1;
    }
    free (semp);
    return 0;
}
